oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/dsl/spec/filter/FilterSpec.java

```diff
@@ -169,7 +169,7 @@ public void sink(@DelegatesTo(SinkSpec.class) final Closure<?> cl) {
         final Optional<AtomicReference<Log>> container = BINDING.get().logContainer();
         if (container.isPresent()) {
             factories.stream()
-                     .map(LogAnalysisListenerFactory::create)
+                     .map(it -> it.create(null))
```

Comment:
I am wondering, why we use `null` here?

Reply:
These two listeners needn't layer param.

Reply:
This listener needn't layer param.

Reply:
@kezhenxu94 Should we separate `LogFilterListener.Factory` and `TrafficAnalysisListener.Factory`/`RecordAnalysisListener.Factory`?

`TrafficAnalysisListener.Factory`/`RecordAnalysisListener.Factory` are more likely `LogSinkListener`.

Reply:
> @kezhenxu94 Should we separate `LogFilterListener.Factory` and `TrafficAnalysisListener.Factory`/`RecordAnalysisListener.Factory`?
> 
> `TrafficAnalysisListener.Factory`/`RecordAnalysisListener.Factory` are more likely `LogSinkListener`.

Hi, I'm not sure, these factories share the same interface and they are used basically in the core log analyzer, so splitting these might bring huge changes.

Reply:
The issue is what I proposed. The create  method is adding layer, but those two for sink are not relative. 

@yswdqz put a null layer for create method. I feel we should change it.

Reply:
@yswdqz Generally tests passed. Please follow this to separate `LogSinkListener` out of `LogAnalysisListener` implementations. And `LogSinkListener#Factory#create` doesn't require layer as it is generally usable.

Reply:
> @yswdqz Generally tests passed. Please follow this to separate `LogSinkListener` out of `LogAnalysisListener` implementations. And `LogSinkListener#Factory#create` doesn't require layer as it is generally usable.

Hi, I’m confused. Currently the reason why there is a `null` in the `create` method is that the 3 factories share the same interface and the interface is used in the core module of log analysis, if we could separate the sink factory we could just remove the null parameter from the create method. Adding a new sink listener factory doesn’t help from what I can tell. Or I misunderstood your request here.

Reply:
@kezhenxu94 The reason is `LogFilterListener`, it is required to have a `layer` as parameter to create `listener`, which we changed in this PR.

But once we add them, we have to change all others due to this hierarchy tree.

<img width="675" alt="image" src="https://user-images.githubusercontent.com/5441976/189624683-9b7d0368-86cb-4d49-9c32-4788e114f1d7.png">

The only reason I asked for this, is to make `create(null)` disappear.

Reply:
> Currently the reason why there is a null in the create method is that the 3 factories share the same interface and the interface is used in the core module of log analysis, if we could separate the sink factory we could just remove the null parameter from the create method.

I feel you are asking the same thing. 3 implementations for one interface. Add one for <2> and <3>, and add one for <1>? I just don't think they are doing the same things, as one for core, the other two for sink. And `null` as a `create` method's parameter seems not right.
I didn't insist on anything relative to which interface and the names. If you want any preference, feel free to propose.

Reply:
> @kezhenxu94 The reason is `LogFilterListener`, it is required to have a `layer` as parameter to create `listener`, which we changed in this PR.
> 
> But once we add them, we have to change all others due to this hierarchy tree.
> 
> <img alt="image" width="675" src="https://user-images.githubusercontent.com/5441976/189624683-9b7d0368-86cb-4d49-9c32-4788e114f1d7.png">
> 
> The only reason I asked for this, is to make `create(null)` disappear.

That’s my point. How to separate the 3 factories?  Add a new sink factory and the add the sink factory here?

https://github.com/apache/skywalking/blob/5ba874a49946571506b6cb3c1abbf0d4ce1863bf/oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/ILogAnalysisListenerManager.java#L23

```yaml
void addSinkListenerFactory(LogSinkListenerFactory factory);
```

Reply:
TrafficAnalysisListener#Factory and RecordAnalysisListener#Factory implement from a new interface, that is all.

Reply:
LogAnalysisListenerFactory interface is changed for core to add a new layer in create method.

Reply:
> > @kezhenxu94 The reason is `LogFilterListener`, it is required to have a `layer` as parameter to create `listener`, which we changed in this PR.
> 
> > 
> 
> > But once we add them, we have to change all others due to this hierarchy tree.
> 
> > 
> 
> > <img alt="image" width="675" src="https://user-images.githubusercontent.com/5441976/189624683-9b7d0368-86cb-4d49-9c32-4788e114f1d7.png">
> 
> > 
> 
> > The only reason I asked for this, is to make `create(null)` disappear.
> 
> 
> 
> That’s my point. How to separate the 3 factories?  Add a new sink factory and then add the sink factory here?
> 
> 
> 
> https://github.com/apache/skywalking/blob/5ba874a49946571506b6cb3c1abbf0d4ce1863bf/oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/ILogAnalysisListenerManager.java#L23
> 
> 
> 
> ```yaml
> 
> void addSinkListenerFactory(LogSinkListenerFactory factory);
> 
> ```
> 
> 
> 
> 

I am not sure whether we need. I remember those two factories are hard coded created.